fix account_objects with invalid marker do not return an error

yinyiqian1 opened the pull request:
fix #4542 
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.

If your branch is on a personal fork and has a name that allows it to
run CI build/test jobs (e.g. "ci/foo"), remember to rename it BEFORE
opening the PR.  This avoids unnecessary redundant test runs. Renaming
the branch after opening the PR will close the PR.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.

For performance-impacting changes, please provide these details:
1. Is this a new feature, bug fix, or improvement to existing functionality?
2. What behavior/functionality does the change impact?
3. In what processing can the impact be measured? Be as specific as possible - e.g. RPC client call, payment transaction that involves LOB, AMM, caching, DB operations, etc.
4. Does this change affect concurrent processing - e.g. does it involve acquiring locks, multi-threaded processing, or async processing?
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

gregtatcam commented on src/xrpld/rpc/detail/RPCHelpers.cpp:
```diff
@@ -250,7 +250,10 @@ getAccountObjects(
-        return mlimit < limit;
+        if (mlimit >= limit)
+            jvResult[jss::account_objects] = Json::arrayValue;
+
```
Please add comments why `true` is always returned in this case.

yinyiqian1 replied:
updated. also added for  `if (mlimit >= limit)`. please check

gregtatcam reviewed (approved):
👍 LGTM

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,104 @@ class AccountObjects_test : public beast::unit_test::suite
+            auto resp = env.rpc("json", "account_objects", to_string(params));
+            auto& accountObjects = resp[jss::result][jss::account_objects];
+            BEAST_EXPECT(!resp[jss::result].isMember(jss::error));
+            BEAST_EXPECT(
```
please also add here check that this is the last page:
```c++
BEAST_EXPECT(!resp[jss::result].isMember(jss::marker));
```

yinyiqian1 replied:
added

Bronek commented on src/xrpld/rpc/handlers/AccountObjects.cpp:
```diff
@@ -283,6 +283,10 @@ doAccountObjects(RPC::JsonContext& context)
             return RPC::invalid_field_error(jss::marker);
     }
 
+    // check if dirIndex is valid
+    if (!dirIndex.isZero() && !ledger->read({ltDIR_NODE, dirIndex}))
+        return RPC::invalid_field_error(jss::marker);
```
since `RPC::getAccountObjects` returning `false` would have the same result, I would move this check into the body of `RPC::getAccountObjects`; this will keep the checks (other than the `marker` format, in the code segment above) in that one function only, which is more robust.

yinyiqian1 replied:
updated

Bronek commented on src/xrpld/rpc/detail/RPCHelpers.cpp:
```diff
@@ -249,8 +249,18 @@ getAccountObjects(
+        if (mlimit >= limit)
+            jvResult[jss::account_objects] = Json::arrayValue;
+
+        // non-zero dirIndex validity was checked in the caller function;
+        // by this point, it should be zero.
```
as commented elsewhere, I would prefer if this function did not rely on the validation of `dirIndex` happening in the caller, but rather move that check into the body of `getAccountObjects`

yinyiqian1 replied:
updated

Bronek reviewed (commented):
Thanks for fixing this bug. I have some comments which would hopefully make this fix a little better.

Bronek commented:
While you are at it, I see that `account_objects` has pretty good unit test coverage but it could be still improved. See the red line 274 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fhandlers%2FAccountObjects.cpp and red line 183 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fdetail%2FRPCHelpers.cpp ; perhaps you could add tests for those two lines as well ?

yinyiqian1 commented:
> While you are at it, I see that `account_objects` has pretty good unit test coverage but it could be still improved. See the red line 274 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fhandlers%2FAccountObjects.cpp and red line 183 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fdetail%2FRPCHelpers.cpp ; perhaps you could add tests for those two lines as well ?

Sure. I just added. But the following line is unreachable because if string s does not have ','   , std::getline will just read to the string end. It wont return error. 
```
if (!std::getline(ss, s, ','))
            return RPC::invalid_field_error(jss::marker);
```

So I also changed it to 
```
if (s.find(',') == std::string::npos)
            return RPC::invalid_field_error(jss::marker);
```

yinyiqian1 commented:
> While you are at it, I see that `account_objects` has pretty good unit test coverage but it could be still improved. See the red line 274 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fhandlers%2FAccountObjects.cpp and red line 183 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fdetail%2FRPCHelpers.cpp ; perhaps you could add tests for those two lines as well ?

Yes. I just added. 

There might be a potential bug here, if we pass in some marker like `xxxxxx,yyyyyy,,,,,,,,1234`, suppose dirIndex and entryIndex part are valid, it will just ignore all the characters after the second `,`. Do you think we should fix this as well?
Also just double check, the marker will always be two parts seperated by `,`, right?
```
std::stringstream ss(marker.asString());
        std::string s;
        if (!std::getline(ss, s, ','))
            return RPC::invalid_field_error(jss::marker);

        if (!dirIndex.parseHex(s))
            return RPC::invalid_field_error(jss::marker);

        if (!std::getline(ss, s, ','))
            return RPC::invalid_field_error(jss::marker);

        if (!entryIndex.parseHex(s))
            return RPC::invalid_field_error(jss::marker);
```

Bronek commented:
> > While you are at it, I see that `account_objects` has pretty good unit test coverage but it could be still improved. See the red line 274 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fhandlers%2FAccountObjects.cpp and red line 183 in https://app.codecov.io/gh/XRPLF/rippled/blob/develop/src%2Fxrpld%2Frpc%2Fdetail%2FRPCHelpers.cpp ; perhaps you could add tests for those two lines as well ?
> 
> Yes. I just added.
> 
> There might be a potential bug here, if we pass in some marker like `xxxxxx,yyyyyy,,,,,,,,1234`, suppose dirIndex and entryIndex part are valid, it will just ignore all the characters after the second `,`. Do you think we should fix this as well? Also just double check, the marker will always be two parts seperated by `,`, right?
> 
> ```
> std::stringstream ss(marker.asString());
>         std::string s;
>         if (!std::getline(ss, s, ','))
>             return RPC::invalid_field_error(jss::marker);
> 
>         if (!dirIndex.parseHex(s))
>             return RPC::invalid_field_error(jss::marker);
> 
>         if (!std::getline(ss, s, ','))
>             return RPC::invalid_field_error(jss::marker);
> 
>         if (!entryIndex.parseHex(s))
>             return RPC::invalid_field_error(jss::marker);
> ```

It would be great to fix it, yes - @mDuo13 any comments ?

Bronek reviewed (approved):
nice !

yinyiqian1 commented:
> It would be great to fix it, yes - @mDuo13 any comments ?

@Bronek  I also fix it in the recent commit. Could you review as well?

Bronek commented on src/xrpld/rpc/handlers/AccountObjects.cpp:
```diff
@@ -268,18 +268,15 @@ doAccountObjects(RPC::JsonContext& context)
-        if (!entryIndex.parseHex(s))
+        if (!entryIndex.parseHex(markerStr.substr(idx + 1)))
             return RPC::invalid_field_error(jss::marker);
     }
```
really nice improvement in parsing here ! :+1:

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,142 @@ class AccountObjects_test : public beast::unit_test::suite
+            std::stringstream ss(marker.asString());
+            std::string s;
+            std::getline(ss, s, ',');
+            s = s + ",0";
```
can you pls also test when this gets corrupted not by `,0` but by a `,` alone ? i.e. valid `dirIndex` followed by a coma, then end of string. Also reverse, i.e. string starts with a coma, followed by valid `entryIndex`, then end of string. Reading up the implementation these both should return the same kind of error, but you need to have unit tests so that the code (which I think is good at this moment) does not degrade due to bad changes in the future.

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,142 @@ class AccountObjects_test : public beast::unit_test::suite
+        // test invalid marker by adding invalid string after the maker.
+        {
+            std::string s = marker.asString();
+            s.append(",1234");
```
I suggest replacing this test with a different one, where `entryIndex` is valid, see my other comment.

yinyiqian1 replied:
This is used to test adding some extra characters after the entryIndex. The marker format is like `xxxx,yyyy,1234`.

Bronek replied:
Nice, I missed it.

Bronek reviewed (approved):
Good change, just some comments to improve unit tests.

Bronek commented:
@yinyiqian1 I allowed myself to merge the current `develop` to your branch, since it was presenting a confusing diff in `API-CHANGELOG.md` . Hope you do not mind.

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,164 @@ class AccountObjects_test : public beast::unit_test::suite
+
+        env.close();
+
+        unsigned const limit = 11;
```
if you make it `static constexpr` ...

Bronek replied:
never mind, you can capture `limit` also without this change. Although `constexpr` would make it more obvious.

yinyiqian1 replied:
can I just update the lambda function capture to `[&]`, you can review in my recent commit.

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,164 @@ class AccountObjects_test : public beast::unit_test::suite
+        auto testInvalidMarker = [&env, &bob](std::string& marker) {
+            Json::Value params;
+            params[jss::account] = bob.human();
+            params[jss::limit] = 11;
```
... then you will be able to use `limit` here, without explicit capture.

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,164 @@ class AccountObjects_test : public beast::unit_test::suite
+        auto const markerStr = marker.asString();
+        auto const& idx = markerStr.find(',');
+        auto const dirIndex = markerStr.substr(0, idx);
+        auto const entryIndex = markerStr.substr(idx + 1);
```
nice !

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,164 @@ class AccountObjects_test : public beast::unit_test::suite
+        // test account_objects with an invalid marker containing invalid
+        // entryIndex with marker: "dirIndex,0"
+        {
+            std::string s = dirIndex + ',';
```
you missed trailing `0` here

yinyiqian1 replied:
Updated in the recent commit. Please review

Bronek commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -1220,6 +1220,164 @@ class AccountObjects_test : public beast::unit_test::suite
+            s.replace(0, 7, "FFFFFFF");
+            s = dirIndex + ',' + s;
+            BEAST_EXPECT(testInvalidMarker(s));
+        }
```
nice test, but can you please move it to follow immediately after the one where you are replacing part of `dirIndex` with `FFFFFFF` ? This way it will be easier to understand how these two are different.

yinyiqian1 replied:
Updated in the recent commit. Please review

Bronek reviewed (approved):
Good change, just minor remarks on unit tests.

Bronek commented:
nice ! @yinyiqian1 do you agree this is ready to merge ?

yinyiqian1 commented:
> nice ! @yinyiqian1 do you agree this is ready to merge ?

Yes. Please merge.

yinyiqian1 commented:
> nice ! @yinyiqian1 do you agree this is ready to merge ?

Please do not merge as I found some issue. will let you know once fixed.

Bronek commented:
@mDuo13 I think this PR does not require change in `API-CHANGELOG.md` because the erroneous behaviour being fixed here is not documented nor makes practical sense to use. You opinion ?

mDuo13 commented:
Yes, this sounds like a bug fix that does not require an API version change.
